Add e2e_latency_millis metric

oguzhanunlu opened the pull request:
This PR adds a new metric `e2e_latency_millis` which measures the difference between an event's `collector_tstamp` and when it was written to BigQuery.

The metric is emitted to statsd only on minutes in which some events are written to BigQuery, i.e. the end of a window.  In other minutes (mid-window) the metric is not defined and not emitted to statsd.

The metric measures the worst-case latency for a window, i.e. latency for the earliest seen `collector_tstamp`.

ref:PDP-1648

istreeter commented on modules/core/src/main/scala/com.snowplowanalytics.snowplow.bigquery/Metrics.scala:
```diff
@@ -35,20 +36,22 @@ object Metrics {
     def initialize[F[_]: Functor](sourceAndAck: SourceAndAck[F]): F[State] =
       sourceAndAck.currentStreamLatency.map { latency =>
-        State(0L, 0L, latency.getOrElse(Duration.Zero))
+        State(0L, 0L, latency.getOrElse(Duration.Zero), Duration.Zero)
```
We should not emit `Duration.Zero` in circumstances when no events are loaded.  Please be consistent with the lake loader.  The default value should be `None` and we should emit nothing when nothing is loaded.

https://github.com/snowplow-incubator/snowplow-lake-loader/blob/0.6.0/modules/core/src/main/scala/com.snowplowanalytics.snowplow.lakes/Metrics.scala#L46

istreeter reviewed (approved):
This looks great.  In particular, very nice work on the tests!

Please can you add a better description to the commit message and PR?  And then I'm happy for you to merge it.  See Lake Loader for a nice example of a commit message.

oguzhanunlu commented:
I just squashed and updated commit message to mimic lake loader [commit](https://github.com/snowplow-incubator/snowplow-lake-loader/commit/ce913a5ca8b67d41cce8522e531ddb2af1e6d637#diff-c936e08d70580ceca5a1f791eda88a2bca5baa0d186a8d2568e6b292c0faa3c2). I also updated the PR description.
